File: packages/sealedrun-py/src/sealedrun/verify.py
"""Verification of a single run (SPEC 13.2)."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from sealedrun.delegation import covers, verify_delegation
from sealedrun.errors import VerificationError
from sealedrun.hashing import digest_size, payload_digest, zero_hash
from sealedrun.keys import KeySet
from sealedrun.schema import validate, validate_extensions
from sealedrun.signing import DOMAIN_RECORD, check_hash, check_signatures
from sealedrun.timeutil import parse_timestamp


@dataclass
class RunReport:
    """What `verify_run` established about a run.

    Attributes:
        first_seq: Sequence number of the first record; above 0 for a partial run.
        last_hash: Hash of the last record, the value a continuation must chain from.
        complete: True when the run ends with `run_end`.
        anchors: Number of anchor records whose reference into the chain was checked.
        anchors_witness_verified: Anchors whose witness proof was checked cryptographically.
            Always 0 in 0.1 (TRUST.md).
        labels_sent_to_cloud: Per data label, the count of non-blocked records with a cloud target.
    """

    run_id: str
    record_count: int
    first_seq: int
    last_hash: str
    complete: bool
    anchors: int = 0
    anchors_witness_verified: int = 0
    labels_sent_to_cloud: dict[str, int] = field(default_factory=dict)


def verify_run(
    records: list[dict[str, Any]],
    delegations: dict[str, dict[str, Any]],
    *,
    payloads: dict[str, bytes] | None = None,
    expected_prev_hash: str | None = None,
) -> RunReport:
    """Verify schema, chain, hashes, delegation and signatures of one run, in record order.

    `delegations` maps `delegation_id` to the delegation object. When `payloads` (digest to body)
    is given, bodies stored in the bundle or inline are checked against their references. A run
    that starts above seq 0 chains from `expected_prev_hash` and takes the first delegation issued
    to its agent, since it has no `run_start` to name one.

    Raises VerificationError at the first failed check.
    """
    if not records:
        raise VerificationError("empty", "run has no records")
    run_id = records[0]["run_id"]
    hash_alg = records[0]["hash_alg"]
    first_seq = records[0]["seq"]
    prev_hash = expected_prev_hash if expected_prev_hash is not None else zero_hash(hash_alg)
    if first_seq == 0 and prev_hash != zero_hash(hash_alg):
        raise VerificationError("chain", "run starts at seq 0 but prev hash is not zero", run_id, 0)

    delegation = _delegation_for_run(records[0], delegations, run_id)
    agent_keys = KeySet.from_json(delegation["agent_keys"])
    report = RunReport(run_id, len(records), first_seq, records[-1]["hash"], False)
    ended = False

    for index, record in enumerate(records):
        seq = first_seq + index
        errors = validate("record.json", record, first_only=True)
        if errors:
            raise VerificationError("schema", "; ".join(errors), run_id, seq)
        ext_errors = validate_extensions(record.get("extensions", {}), first_only=True)
        if ext_errors:
            raise VerificationError("schema", "; ".join(ext_errors), run_id, seq)
        if record["run_id"] != run_id:
            raise VerificationError("run", "record belongs to another run", run_id, seq)
        if record["seq"] != seq:
            raise VerificationError("seq", f"expected seq {seq}, got {record['seq']}", run_id, seq)
        if record["hash_alg"] != hash_alg:
            raise VerificationError("hash_alg", "hash algorithm changed within run", run_id, seq)
        if len(record["hash"]) != digest_size(hash_alg) * 2:
            raise VerificationError("hash_alg", "hash length does not match algorithm", run_id, seq)
        if ended:
            raise VerificationError("run_end", "record after run_end", run_id, seq)
        if record["prev_hash"] != prev_hash:
            raise VerificationError(
                "chain", "prev_hash does not match previous record", run_id, seq
            )
        if not check_hash(record):
            raise VerificationError("hash", "record hash mismatch", run_id, seq)
        if record["agent_id"] != delegation["agent_id"]:
            raise VerificationError("delegation", "agent_id differs from delegation", run_id, seq)
        if record["principal_id"] != delegation["principal_id"]:
            raise VerificationError(
                "delegation", "principal_id differs from delegation", run_id, seq
            )
        if not covers(delegation, parse_timestamp(record["occurred_at"])):
            raise VerificationError("delegation", "record outside delegation validity", run_id, seq)
        if not check_signatures(record, DOMAIN_RECORD, agent_keys):
            raise VerificationError("signature", "agent signature invalid", run_id, seq)
        if payloads is not None:
            _check_payload(record, payloads, hash_alg, run_id, seq)
        if record["kind"] == "anchor":
            _check_anchor(record, records, first_seq, run_id, seq)
            report.anchors += 1
        if record["kind"] == "run_end":
            ended = True
        if record["target"].get("location") == "cloud" and record.get("outcome") != "blocked":
            for label in record["data_labels"]:
                report.labels_sent_to_cloud[label] = report.labels_sent_to_cloud.get(label, 0) + 1
        prev_hash = record["hash"]

    report.complete = ended
    return report


def _delegation_for_run(
    first: dict[str, Any], delegations: dict[str, dict[str, Any]], run_id: str
) -> dict[str, Any]:
    binding = first.get("extensions", {}).get("sealedrun.delegation")
    if first["seq"] == 0:
        if not binding:
            raise VerificationError("delegation", "run_start lacks sealedrun.delegation", run_id, 0)
        delegation = delegations.get(binding["delegation_id"])
        if delegation is None:
            raise VerificationError("delegation", "delegation not in bundle", run_id, 0)
        if delegation["hash"] != binding["hash"]:
            raise VerificationError("delegation", "delegation hash mismatch", run_id, 0)
    else:
        candidates = [d for d in delegations.values() if d["agent_id"] == first["agent_id"]]
        if not candidates:
            raise VerificationError("delegation", "no delegation for agent", run_id, first["seq"])
        delegation = candidates[0]
    problem = verify_delegation(delegation)
    if problem:
        raise VerificationError("delegation", problem, run_id, first["seq"])
    return delegation


def _check_payload(
    record: dict[str, Any], payloads: dict[str, bytes], hash_alg: str, run_id: str, seq: int
) -> None:
    ref = record.get("payload")
    if not ref or ref["storage"] not in ("bundle", "inline"):
        return
    for side in ("request", "response"):
        expected = ref.get(f"{side}_hash")
        if expected is None:
            continue
        body = payloads.get(expected)
        if body is None:
            raise VerificationError("payload", f"{side} body missing", run_id, seq)
        if payload_digest(hash_alg, body) != expected or len(body) != ref[f"{side}_size"]:
            raise VerificationError("payload", f"{side} body digest mismatch", run_id, seq)


def _check_anchor(
    record: dict[str, Any], records: list[dict[str, Any]], first_seq: int, run_id: str, seq: int
) -> None:
    """Check that an anchor points at an earlier record of this run and carries its hash.

    SPEC 8.1: the receipt digest is the digest the witness was given, so it must equal
    `anchored_hash`. The witness signature is not checked in 0.1.
    """
    anchor = record["extensions"]["sealedrun.anchor"]
    index = anchor["anchored_seq"] - first_seq
    if index < 0 or index >= len(records) or anchor["anchored_seq"] >= seq:
        raise VerificationError("anchor", "anchored_seq not in run before anchor", run_id, seq)
    if records[index]["hash"] != anchor["anchored_hash"]:
        raise VerificationError("anchor", "anchored_hash does not match record", run_id, seq)
    if anchor["receipt"]["digest"] != anchor["anchored_hash"]:
        raise VerificationError(
            "anchor", "receipt digest does not match anchored_hash", run_id, seq
        )

File: packages/sealedrun-ts/src/verify.ts
import { covers, verifyDelegation } from "./delegation.js";
import { VerificationError } from "./errors.js";
import { digestSize, type HashAlg, isHashAlg, payloadDigest, zeroHash } from "./hashing.js";
import { assertRecord } from "./structure.js";
import { checkHash, checkSignatures, DOMAIN_RECORD } from "./signing.js";
import { parseTimestamp } from "./time.js";
import type { SealedRunRecord, Delegation } from "./types.js";

/** Result of a successful {@link verifyRun}. */
export interface RunReport {
  runId: string;
  /** Records in the verified slice, which may be less than the whole run. */
  recordCount: number;
  firstSeq: number;
  lastHash: string;
  /** True when the slice contains a `run_end` record. */
  complete: boolean;
  /** Anchor records whose `anchored_hash` and receipt digest matched the run. */
  anchors: number;
  /** Anchors whose witness proof was checked cryptographically. Always 0 in 0.1 (TRUST.md). */
  anchorsWitnessVerified: number;
  /** Per data label, the number of non-blocked records whose target location is `cloud`. */
  labelsSentToCloud: Record<string, number>;
}

/** Optional inputs of {@link verifyRun}. */
export interface VerifyRunOptions {
  /**
   * Bodies by base64url digest. When given, the bodies of `bundle` and `inline` payloads must be
   * present and match their digest and size.
   */
  payloads?: Map<string, Uint8Array>;
  /** `prev_hash` the first record must carry when the slice does not start at seq 0. */
  expectedPrevHash?: string;
}

/**
 * Verifies one run or a contiguous slice of it: structure, seq contiguity, hash chain, record
 * hashes, delegation binding and validity, agent signatures, anchors, and nothing after `run_end`
 * (SPEC 13.2 steps 3 to 5).
 *
 * @remarks
 * All records are checked against a single Delegation. For a run that starts at seq 0 it is the
 * one bound by `sealedrun.delegation` in `run_start` (SPEC 6.4). For a slice it is the first
 * Delegation in the map issued to the agent.
 *
 * @param records - Records in ascending seq order.
 * @param delegations - Delegations by `delegation_id`.
 * @throws VerificationError naming the first failed check, with run id and seq.
 * @throws TypeError if a timestamp is not in the SPEC 3 form.
 */
export function verifyRun(
  records: SealedRunRecord[],
  delegations: Map<string, Delegation>,
  options: VerifyRunOptions = {},
): RunReport {
  const first = records[0];
  if (!first) throw new VerificationError("empty", "run has no records");
  const runId = first.run_id;
  if (!isHashAlg(first.hash_alg)) {
    throw new VerificationError("hash_alg", "unsupported hash algorithm", runId, first.seq);
  }
  const hashAlg: HashAlg = first.hash_alg;
  const firstSeq = first.seq;
  let prevHash = options.expectedPrevHash ?? zeroHash(hashAlg);
  if (firstSeq === 0 && prevHash !== zeroHash(hashAlg)) {
    throw new VerificationError("chain", "run starts at seq 0 but prev hash is not zero", runId, 0);
  }
  const delegation = delegationForRun(first, delegations, runId);
  const last = records[records.length - 1] ?? first;
  const report: RunReport = {
    runId,
    recordCount: records.length,
    firstSeq,
    lastHash: last.hash,
    complete: false,
    anchors: 0,
    anchorsWitnessVerified: 0,
    labelsSentToCloud: {},
  };
  let ended = false;

  records.forEach((record, index) => {
    const seq = firstSeq + index;
    try {
      assertRecord(record);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new VerificationError("schema", message, runId, seq);
    }
    if (record.run_id !== runId) {
      throw new VerificationError("run", "record belongs to another run", runId, seq);
    }
    if (record.seq !== seq) {
      throw new VerificationError("seq", `expected seq ${seq}, got ${record.seq}`, runId, seq);
    }
    if (record.hash_alg !== hashAlg) {
      throw new VerificationError("hash_alg", "hash algorithm changed within run", runId, seq);
    }
    if (record.hash.length !== digestSize(hashAlg) * 2) {
      throw new VerificationError("hash_alg", "hash length does not match algorithm", runId, seq);
    }
    if (ended) throw new VerificationError("run_end", "record after run_end", runId, seq);
    if (record.prev_hash !== prevHash) {
      throw new VerificationError("chain", "prev_hash does not match previous record", runId, seq);
    }
    if (!checkHash(record)) throw new VerificationError("hash", "record hash mismatch", runId, seq);
    if (record.agent_id !== delegation.agent_id) {
      throw new VerificationError("delegation", "agent_id differs from delegation", runId, seq);
    }
    if (record.principal_id !== delegation.principal_id) {
      throw new VerificationError("delegation", "principal_id differs from delegation", runId, seq);
    }
    if (!covers(delegation, parseTimestamp(record.occurred_at))) {
      throw new VerificationError("delegation", "record outside delegation validity", runId, seq);
    }
    if (!checkSignatures(record, DOMAIN_RECORD, delegation.agent_keys)) {
      throw new VerificationError("signature", "agent signature invalid", runId, seq);
    }
    if (options.payloads) checkPayload(record, options.payloads, hashAlg, runId, seq);
    if (record.kind === "anchor") {
      checkAnchor(record, records, firstSeq, runId, seq);
      report.anchors += 1;
    }
    if (record.kind === "run_end") ended = true;
    if (record.target.location === "cloud" && record.outcome !== "blocked") {
      for (const label of record.data_labels) {
        report.labelsSentToCloud[label] = (report.labelsSentToCloud[label] ?? 0) + 1;
      }
    }
    prevHash = record.hash;
  });

  report.complete = ended;
  return report;
}

function delegationForRun(
  first: SealedRunRecord,
  delegations: Map<string, Delegation>,
  runId: string,
): Delegation {
  let delegation: Delegation | undefined;
  if (first.seq === 0) {
    const binding = first.extensions?.["sealedrun.delegation"];
    if (!binding)
      throw new VerificationError("delegation", "run_start lacks sealedrun.delegation", runId, 0);
    delegation = delegations.get(String(binding["delegation_id"]));
    if (!delegation)
      throw new VerificationError("delegation", "delegation not in bundle", runId, 0);
    if (delegation.hash !== binding["hash"]) {
      throw new VerificationError("delegation", "delegation hash mismatch", runId, 0);
    }
  } else {
    delegation = [...delegations.values()].find((d) => d.agent_id === first.agent_id);
    if (!delegation) {
      throw new VerificationError("delegation", "no delegation for agent", runId, first.seq);
    }
  }
  const problem = verifyDelegation(delegation);
  if (problem) throw new VerificationError("delegation", problem, runId, first.seq);
  return delegation;
}

function checkPayload(
  record: SealedRunRecord,
  payloads: Map<string, Uint8Array>,
  hashAlg: HashAlg,
  runId: string,
  seq: number,
): void {
  const ref = record.payload;
  if (!ref || (ref.storage !== "bundle" && ref.storage !== "inline")) return;
  for (const side of ["request", "response"] as const) {
    const expected = ref[`${side}_hash`];
    if (expected === undefined) continue;
    const body = payloads.get(expected);
    if (!body) throw new VerificationError("payload", `${side} body missing`, runId, seq);
    if (payloadDigest(hashAlg, body) !== expected || body.length !== ref[`${side}_size`]) {
      throw new VerificationError("payload", `${side} body digest mismatch`, runId, seq);
    }
  }
}

/**
 * Checks that an anchor record points at an earlier record of the same run and that its receipt
 * carries the digest the witness was given (SPEC 8.1).
 *
 * @remarks
 * The witness signature is not checked in 0.1.
 */
function checkAnchor(
  record: SealedRunRecord,
  records: SealedRunRecord[],
  firstSeq: number,
  runId: string,
  seq: number,
): void {
  const anchor = record.extensions?.["sealedrun.anchor"];
  if (!anchor)
    throw new VerificationError("anchor", "anchor record lacks sealedrun.anchor", runId, seq);
  const anchoredSeq = anchor["anchored_seq"];
  const anchoredHash = anchor["anchored_hash"];
  if (!Number.isSafeInteger(anchoredSeq) || (anchoredSeq as number) < 0) {
    throw new VerificationError("anchor", "anchored_seq is not a non-negative integer", runId, seq);
  }
  if (typeof anchoredHash !== "string") {
    throw new VerificationError("anchor", "anchored_hash is missing", runId, seq);
  }
  const index = (anchoredSeq as number) - firstSeq;
  if (index < 0 || index >= records.length || (anchoredSeq as number) >= seq) {
    throw new VerificationError("anchor", "anchored_seq not in run before anchor", runId, seq);
  }
  if (records[index]?.hash !== anchoredHash) {
    throw new VerificationError("anchor", "anchored_hash does not match record", runId, seq);
  }
  const receipt = anchor["receipt"] as Record<string, unknown>;
  if (receipt["digest"] !== anchoredHash) {
    throw new VerificationError(
      "anchor",
      "receipt digest does not match anchored_hash",
      runId,
      seq,
    );
  }
}
